### PR TITLE
Build applications in sub-directory

### PR DESCRIPTION
**Pull request by guilleiguaran**
For details see #3, elixir-lang/elixir#1046, elixir-lang/dynamo#101, elixir-lang/dynamo#99

### CODE REVIEW

File: bin/compile [hunk 1]
```diff
@@ -33,6 +35,11 @@ ex_tarball=${ex_ver}.tar
 ELIXIR_TARBALL_URL="https://github.com/elixir-lang/elixir/tarball/${ex_ver}"
 EXROOT=${build}/ex
 
+mkdir ${APP_PATH}
+for file in ${build}/*; do
```
Comment (guilleiguaran) [inline bin/compile]: This is ignoring hidden files, any ideas about how to make it work also for hidden files?

File: bin/compile [hunk 2]
```diff
@@ -15,6 +15,8 @@ test -z ${build} && exit
 cache=$(cd "$2/" && pwd)
 test -z ${cache} && exit
 
+APP_NAME=$(grep "app: \(:\)" ${build}/mix.exs | cut -d ":" -f 3 | cut -d "," -f 1)
```
Comment (guilleiguaran) [inline bin/compile:18]: Someone have a better idea about how to extract application name?

File: bin/compile (hunk 1 above)
Comment (goshacmd) [inline bin/compile]: Yeah, `for file in ${build}/* ${build}/.*` should do.

**Comment by guilleiguaran** [comment]
ok, with my test app (named dynamita) I'm getting this file list in heroku:

``` elixir
{:ok, ["rebar", "dynamita", "otp", "ex", "Procfile"]}
```

@goshakkk I still need to figure out how to run the app that is in the subfolder (I think heroku runs all the commands specified in the Procfile in `/app` folder), any ideas?

/cc @josevalim

**Comment by goshacmd** [comment]
@guilleiguaran probably the build pack could create a [.profile.d](https://devcenter.heroku.com/articles/profiled) script which would prepend app's dir to PATH/cd to app's dir.

**Comment by guilleiguaran** [comment]
@goshakkk thanks!! I think I can use .profile.d and export the app folder in `MIX_PATH` to do it!!!

**Comment by guilleiguaran** [comment]
Not sure about what's wrong, `MIX_PATH` is set correctly (I think :grin:):

``` elixir
iex(1)> System.get_env "MIX_PATH"
"/app/dynamita/ebin"
```

but in Heroku logs I'm getting:

```
2013-07-24T07:58:16.654454+00:00 heroku[web.1]: Starting process with command `MIX_ENV=prod mix server -p 45724`
2013-07-24T07:58:20.552768+00:00 app[web.1]: ** (Mix) The task server could not be found
```

/cc @josevalim

**Comment by guilleiguaran** [comment]
nvm, looks like I need to append `/app/dynamita/deps/dynamo/ebin` to MIX_PATH

**Comment by guilleiguaran** [comment]
well, looks like I have a different error now:

```
2013-07-24T08:20:36.581606+00:00 heroku[web.1]: Starting process with command `MIX_ENV=prod mix server -p 14675`
2013-07-24T08:20:39.183828+00:00 app[web.1]: ** (ArgumentError) argument error
2013-07-24T08:20:39.215477+00:00 app[web.1]:     :erlang.length(nil)
2013-07-24T08:20:39.215477+00:00 app[web.1]:     /tmp/build_15el1thjh0ual/ex/lib/mix/lib/mix/cli.ex:38: Mix.CLI.run_task/2
2013-07-24T08:20:39.215477+00:00 app[web.1]:     /tmp/build_15el1thjh0ual/ex/lib/elixir/lib/code.ex:281: Code.require_file/2
2013-07-24T08:20:39.215477+00:00 app[web.1]:     /tmp/build_15el1thjh0ual/dynamita/deps/dynamo/lib/mix/tasks/server.ex:22: Mix.Tasks.Server.run/1
2013-07-24T08:20:39.215477+00:00 app[web.1]:     /tmp/build_15el1thjh0ual/ex/lib/elixir/lib/kernel/cli.ex:305: Kernel.CLI.process_command/2
```

ideas?

**Comment by josevalim** [comment]
Changing `MIX_PATH` like that is definitely not going to work.

What needs to be done is to simply change the Procfile to start the server inside the dynamita directory. According to @hone, everything should just work as long as the Procfile is in /app/Procfile and it knows how to start the server properly.

**Comment by hone** [comment]
Let me know if I can help here.

**Comment by guilleiguaran** [comment]
@hone thanks!! I'll ping you if I need any help

@josevalim then I guess the right solution in this case is change Procfile to `cd dynamita &&  MIX_ENV=prod mix server -p $PORT`, right?

**Comment by goshacmd** [comment]
@guilleiguaran but having to cd in Procfile seems like a weird solution. As I suggested, maybe trying to cd in .profile.d would do?

(Also, should the buildpack perhaps set `MIX_ENV` to `prod` so the users don't have to before every Procfile line?)

**Comment by Tonkpils** [comment]
@goshakkk in that case wouldn't it make sense to create the Procfile in the buildpack itself?

**Comment by guilleiguaran** [comment]
@goshakkk thanks!!! It works fine

**Comment by goshacmd** [comment]
@guilleiguaran awesome! Can it be merged already?

**Comment by guilleiguaran** [comment]
@goshakkk Yes, it's finished :smile:

**Comment by goshacmd** [comment]
Grrrreat! :sparkles: Thank you.

**Pull request merged**